## Abhishek Sharma
### CS 2nd Year | Section : "I" | Roll No.: 01

### Enrollment No.: 12019009001127

## Advanced Artificial Intelligence and Machine Learning Lab 7

#### Faculty : PB, MB, BM | Date : 06.04.2021

### Decision Tree Classifier on Digits Dataset

In [1]:
from sklearn.tree import DecisionTreeClassifier

In [2]:
from sklearn.datasets import load_digits

In [3]:
digits = load_digits()

In [4]:
from sklearn.model_selection import train_test_split

In [5]:
x_train, x_test, y_train, y_test = train_test_split(digits.data,digits.target, test_size = 0.2)

In [6]:
dtc = DecisionTreeClassifier()

In [7]:
dtc.fit(x_train, y_train)

DecisionTreeClassifier()

In [8]:
dtc.score(x_test, y_test)

0.8444444444444444

Using the Max_depth of 20 in the Classifier

In [9]:
dtc = DecisionTreeClassifier(max_depth = 20)

In [10]:
dtc.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=20)

In [11]:
dtc.score(x_test, y_test)

0.8777777777777778

Using the max_depth of 5 in the classifier

In [12]:
dtc = DecisionTreeClassifier(max_depth = 5)

In [13]:
dtc.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=5)

In [14]:
dtc.score(x_test, y_test)

0.6305555555555555

Using the max_depth of 50 in the classifier

In [15]:
dtc = DecisionTreeClassifier(max_depth = 50)

In [16]:
dtc.fit(x_train, y_train)

DecisionTreeClassifier(max_depth=50)

In [17]:
dtc.score(x_test, y_test)

0.8444444444444444

### Random Forest Classifier on Digits Dataset

In [18]:
from sklearn.ensemble import RandomForestClassifier

In [19]:
rfc = RandomForestClassifier()

In [20]:
rfc.fit(x_train, y_train)

RandomForestClassifier()

In [21]:
rfc.score(x_test, y_test)

0.9805555555555555

### Decision Tree Regressor on the Boston Dataset

In [22]:
from sklearn.datasets import load_boston

In [23]:
import numpy as np
import pandas as pd

In [24]:
boston = load_boston()

In [25]:
data = pd.DataFrame(boston.data, columns = boston.feature_names)

In [26]:
data['MEDV'] = pd.DataFrame(boston.target) 

In [27]:
x = data[['RM','ZN', 'LSTAT']]

In [28]:
y = data['MEDV']

In [29]:
from sklearn.tree import DecisionTreeRegressor

In [30]:
dt = DecisionTreeRegressor()

In [31]:
x_train1, x_test1, y_train1, y_test1 = train_test_split(x,y, test_size = 0.2)

In [32]:
dt.fit(x_train1, y_train1)

DecisionTreeRegressor()

In [33]:
yTestPredict = dt.predict(x_test1)

In [34]:
from sklearn.metrics import mean_squared_error
np.sqrt(mean_squared_error(y_test1, yTestPredict))

5.661834733460489

### Random Forest Regressor on Boston Dataset

In [58]:
from sklearn.ensemble import RandomForestRegressor

In [61]:
rfr = RandomForestRegressor()

In [62]:
rfr.fit(x_train1, y_train1)

RandomForestRegressor()

In [63]:
yTestPredict = rfr.predict(x_test1)

In [64]:
np.sqrt(mean_squared_error(y_test1, yTestPredict))

4.8785307043548825

### Bagging Classifier on Digits Dataset based on Decision Tree Classifier

In [40]:
from sklearn.ensemble import BaggingClassifier

In [49]:
bag = BaggingClassifier(DecisionTreeClassifier())

In [50]:
bag.fit(x_train, y_train)

BaggingClassifier(base_estimator=DecisionTreeClassifier())

In [51]:
bag.score(x_test, y_test)

0.9194444444444444

### Bagging Regressor on Boston Dataset based on Decision Tree Regressor

In [44]:
from sklearn.ensemble import BaggingRegressor

In [52]:
bgr = BaggingRegressor(DecisionTreeRegressor())

In [53]:
bgr.fit(x_train1, y_train1)

BaggingRegressor(base_estimator=DecisionTreeRegressor())

In [54]:
yTestPredict = bgr.predict(x_test1)

In [55]:
np.sqrt(mean_squared_error(y_test1, yTestPredict))

5.003237089380137

### Bagging Regressor on Boston Dataset based on Random Forest Regressor

In [65]:
bgr = BaggingRegressor(RandomForestRegressor())

In [66]:
bgr.fit(x_train1, y_train1)

BaggingRegressor(base_estimator=RandomForestRegressor())

In [67]:
yTestPredict = bgr.predict(x_test1)

In [68]:
np.sqrt(mean_squared_error(y_test1, yTestPredict))

4.589384553521099

### ADA Boost Classifier on Digits Dataset based on decision tree classifer

In [69]:
from sklearn.ensemble import AdaBoostClassifier

In [71]:
ada = AdaBoostClassifier(DecisionTreeClassifier())

In [72]:
ada.fit(x_train, y_train)

AdaBoostClassifier(base_estimator=DecisionTreeClassifier())

In [73]:
ada.score(x_test, y_test)

0.8527777777777777

### Logistic Regression

In [74]:
from sklearn.linear_model import LogisticRegression

In [75]:
reg = LogisticRegression(max_iter = 5000)

In [76]:
reg.fit(x_train, y_train)

LogisticRegression(max_iter=5000)

In [77]:
reg.score(x_test, y_test)

0.9527777777777777

### ADA Boost Classifier on Digits Dataset based on Logistic regression

In [78]:
ada = AdaBoostClassifier(LogisticRegression(max_iter = 5000))

In [79]:
ada.fit(x_train, y_train)

AdaBoostClassifier(base_estimator=LogisticRegression(max_iter=5000))

In [81]:
ada.score(x_test, y_test)

0.9138888888888889

### Conclusion :
After the analysis of the regressor and classifier models on the Digits and boston Dataset respectively. We have conclude the scores of the each models.

**Based on the model fit score of Classifier models we have created the table :**

|Classifier Model Name | Score|
|:---:|:---:|
|Decision Tree Classifier | 0.84 |
|Decision Tree Classifier (max_depth = 20)| 0.88 |
|Decision Tree Classifier (max_depth = 5)|0.63|
|Decision Tree Classifier (max_depth = 50)|0.84|
|Random Forest Classifier|0.98|
|Bagging Classifier(base_estimator=DecisionTreeClassifier())|0.92|
|Ada Boost Classifier(base_estimator=DecisionTreeClassifier())|0.85|
|Ada Boost Classifier(base_estimator=LogisticRegression(max_iter=5000))|0.91|

**Based on the mean squared error of Regressor models we have created the table :**

|Regressor Model Name | Mean squared error|
|:---:|:---:|
|Decision Tree Regressor | 5.66 |
|Random Forest Regressor | 4.87 |
|Bagging Regressor(base_estimator=DecisionTreeRegressor())| 5.01|
|Bagging Regressor(base_estimator=RandomForestRegressor())|4.59|


*******************************************
### <div align = 'center'> Thank You! </div>